src/BlockTypes/AddToCartForm.php

```diff
@@ -58,10 +50,460 @@ protected function render( $attributes, $content, $block ) {
 			esc_attr( $classes_and_styles['classes'] ),
 			esc_attr( $classname ),
 			esc_attr( $classes_and_styles['styles'] ),
-			$product
+			$add_to_cart_form
 		);
 	}
 
+	/**
+	 * Return the Add to Cart Form for a given product.
+	 *
+	 * @param \WC_Product $product Product object.
+	 *
+	 * @return string
+	 */
+	protected function add_to_cart_form( $product ) {
+		$product_type = $product->get_type();
+
+		switch ( $product_type ) {
+			case 'simple':
+				$add_to_cart_form = $this->add_simple_product_to_cart( $product );
+				break;
+			case 'variable':
+				$add_to_cart_form = $this->add_variable_product_to_cart( $product );
+				break;
+			case 'grouped':
+				$add_to_cart_form = $this->add_grouped_product_to_cart( $product );
+				break;
+			case 'external':
+				$add_to_cart_form = $this->add_external_product_to_cart( $product );
+				break;
+			default:
+				$add_to_cart_form = '';
+				break;
+		}
+
+		ob_start();
+
+		/**
+		 * Hook: woocommerce_before_add_to_cart_form.
+		 *
+		 * @since TBD
+		 */
+		do_action( 'woocommerce_before_add_to_cart_form' );
+
+		echo $add_to_cart_form; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+
+		/**
+		 * Hook: woocommerce_after_add_to_cart_form.
+		 *
+		 * @since TBD
+		 */
+		do_action( 'woocommerce_after_add_to_cart_form' );
+
+		return ob_get_clean();
+	}
+
+	/**
+	 * The add to cart form for a simple product.
+	 *
+	 * @param \WC_Product $product Product object.
+	 */
+	protected function add_simple_product_to_cart( $product ) {
+		if ( ! $product->is_purchasable() || ! $product->is_in_stock() ) {
+			return '';
+		}
+
+		ob_start();
+		echo wc_get_stock_html( $product ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+		?>
+		<form class="cart" action="<?php esc_url( $product->get_permalink() ); ?>" method="post" enctype="multipart/form-data">
+			<?php
+			$this->get_quantity_input( $product );
+			$this->add_to_cart_button( $product, 'add-to-cart' );
+			?>
+		</form>
+		<?php
+
+		return ob_get_clean();
+	}
+
+	/**
+	 * The add to cart form for a variable product.
+	 *
+	 * @param \WC_Product $product Product object.
+	 *
+	 * @return string
+	 */
+	protected function add_variable_product_to_cart( $product ) {
+		wp_enqueue_script( 'wc-add-to-cart-variation' );
+		$attributes           = $product->get_variation_attributes();
+		$attribute_keys       = array_keys( $attributes );
+		$available_variations = $product->get_available_variations();
+		$variations_json      = wp_json_encode( $available_variations );
+		$variations_attr      = wc_esc_json( $variations_json );
```

Comment:
`$variations_json` isn't being used separately in `AddToCartForm::variable_product_form` so I'd remove the unnecessary extra variable here and in the function signature.
```suggestion
		$variations_attr      = wc_esc_json( wp_json_encode( $variations_json ) );
```

Reply:
Makes sense: done ✅ !